clients/csrf/csrf_token_manager_impl.go

```diff
@@ -0,0 +1,77 @@
+package csrf
+
+import (
+	"net/http"
+)
+
+const XCsrfHeader = "X-Csrf-Header"
+const XCsrfToken = "X-Csrf-Token"
+const CsrfTokenHeaderRequiredValue = "Required"
+
+type CsrfTokenManagerImpl struct {
+	request          *http.Request
+	transport        *Transport
+	csrfTokenFetcher CsrfTokenFetcher
+}
+
+func NewCsrfTokenManagerImpl(transport *Transport, request *http.Request, csrfTokenFetcher CsrfTokenFetcher) *CsrfTokenManagerImpl {
+	return &CsrfTokenManagerImpl{request: request, transport: transport, csrfTokenFetcher: csrfTokenFetcher}
+}
+
+func (c *CsrfTokenManagerImpl) setCsrfToken() error {
+	if c.request == nil || !c.isProtectionRequired(c.request, c.transport) {
+		return nil
+	}
+	err := c.initializeToken(false, getFetchNewTokenUrl(c.request))
+	if err != nil {
+		return err
+	}
+
+	c.updateCurrentCsrfTokens(c.request, c.transport)
+
+	return nil
+}
+
+func (c *CsrfTokenManagerImpl) initializeToken(force bool, url string) error {
+	if force || !c.transport.Csrf.IsInitialized {
+		var err error
+		c.transport.Csrf.Header, c.transport.Csrf.Token, err = c.csrfTokenFetcher.FetchNewCsrfToken(url, c.request)
+		if err != nil {
+			return err
+		}
+		c.transport.Csrf.IsInitialized = true
```

Comment:
I am not happy with such "global" state variables

Reply:
There is no other way to preserve the initial state.